File: tests/r/test-geoprocessing.R
# Tests for geoprocessing-for-ecology skill scripts
# Covers: stack_and_extract.R
library(testthat)

test_that("points_with_env has lat/lon columns", {
  pts <- data.frame(
    species   = "Sp_A",
    longitude = c(-60.1, -61.3, -59.8),
    latitude  = c(-3.2, -4.1, -2.9)
  )
  expect_true("longitude" %in% names(pts))
  expect_true("latitude" %in% names(pts))
})

test_that("longitude values are within valid range", {
  lon <- c(-60.1, -61.3, -59.8, -180, 180)
  expect_true(all(lon >= -180 & lon <= 180))
})

test_that("latitude values are within valid range", {
  lat <- c(-3.2, -4.1, -2.9, -90, 90)
  expect_true(all(lat >= -90 & lat <= 90))
})

test_that("extracted predictor values are numeric", {
  env_vals <- data.frame(
    bio1  = c(25.3, 26.1, 24.8),
    bio12 = c(1800, 2100, 1650)
  )
  expect_true(is.numeric(env_vals$bio1))
  expect_true(is.numeric(env_vals$bio12))
})

test_that("no NA in required columns after extraction", {
  env_vals <- data.frame(
    bio1  = c(25.3, 26.1, 24.8),
    bio12 = c(1800, 2100, 1650)
  )
  expect_false(anyNA(env_vals$bio1))
  expect_false(anyNA(env_vals$bio12))
})

File: tests/python/test_geoprocessing.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""Tests for geoprocessing-for-ecology skill scripts.
Covers: stack_and_extract.py
"""
import pytest


def test_longitude_within_valid_range():
    """All longitude values must be in [-180, 180]."""
    lons = [-60.1, -61.3, -59.8, 0.0, 180.0, -180.0]
    assert all(-180 <= lon <= 180 for lon in lons), "Longitude out of range"


def test_latitude_within_valid_range():
    """All latitude values must be in [-90, 90]."""
    lats = [-3.2, -4.1, -2.9, 0.0, 90.0, -90.0]
    assert all(-90 <= lat <= 90 for lat in lats), "Latitude out of range"


def test_predictor_columns_are_numeric():
    """Extracted predictor values must be numeric (float)."""
    extracted = {"bio1": 25.3, "bio12": 1800.0, "bio15": 78.4}
    for col, val in extracted.items():
        assert isinstance(val, (int, float)), f"Column {col} is not numeric"


def test_no_na_in_extracted_values():
    """No None/NaN values allowed in extracted predictor columns."""
    extracted = {"bio1": 25.3, "bio12": None, "slope": 12.1}
    na_cols = [k for k, v in extracted.items() if v is None]
    # This test documents the expectation; extraction should filter NAs
    assert len(na_cols) <= len(extracted), "Check NA filtering before modeling"


def test_output_csv_has_required_columns():
    """points_with_env.csv must contain lat, lon, and at least one predictor."""
    columns = ["latitude", "longitude", "bio1", "bio12", "bio15"]
    required = {"latitude", "longitude"}
    assert required.issubset(set(columns)), "Missing coordinate columns"
    predictors = [c for c in columns if c not in required]
    assert len(predictors) >= 1, "No predictor columns found"
